Automatically skip native builds if not needed

hawkrives opened the pull request:
Closes #2582 

```
[22:00:03]: -------------------------------------------------
[22:00:03]: --- Step: git diff --name-only 'master' 'foo' ---
[22:00:03]: -------------------------------------------------
[22:00:03]: $ git diff --name-only 'master' 'foo'
[22:00:03]: ▸ .circleci/config.yml
[22:00:03]: should_build? some native file changed, so yes
```

(this PR should trigger a native build; submitting early to test.)

---

With this PR, `fastlane` will automatically skip native iOS/Android builds when only JS code has changed, as it can't really affect the native results.

So we check if any of these files have changed:

```ruby
  source_globs = [
    '.circleci/**',
    'e2e/**',
    'fastlane/**',
    'scripts/**',
    'Gemfile.lock',
  ]

  case lane_context[:PLATFORM_NAME]
  when :android
    platform_globs = [
      'android/**',
      '{modules,source}/**/*.java',
    ]
  when :ios
    platform_globs = [
      'ios/**',
      '{modules,source}/**/*.{h,m,mm}',
    ]
  end
```

_or_, if any package.json dependencies/devDependencies that include the word "react" or "jsc" have changed.

hawkrives committed:
invoke should_build? in auto_build

hawkrives committed:
add code for determining if native files changed

hawkrives committed:
add should_build? check in api_keys_available?

hawkrives committed:
reorganize auto_beta to simplify the conditions

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+  end
+
+  source_globs = [
+    ".circleci/**",
```
```suggestion
    '.circleci/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+
+  source_globs = [
+    ".circleci/**",
+    "e2e/**",
```
```suggestion
    'e2e/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+  source_globs = [
+    ".circleci/**",
+    "e2e/**",
+    "fastlane/**",
```
```suggestion
    'fastlane/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+    ".circleci/**",
+    "e2e/**",
+    "fastlane/**",
+    "scripts/**",
```
```suggestion
    'scripts/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+    "e2e/**",
+    "fastlane/**",
+    "scripts/**",
+    "Gemfile.lock",
```
```suggestion
    'Gemfile.lock',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+  case lane_context[:PLATFORM_NAME]
+  when :android
+    platform_globs = [
+      "android/**",
```
```suggestion
      'android/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+  when :android
+    platform_globs = [
+      "android/**",
+      "{modules,source}/**/*.java",
```
```suggestion
      '{modules,source}/**/*.java',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+    ]
+  when :ios
+    platform_globs = [
+      "ios/**",
```
```suggestion
      'ios/**',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+  when :ios
+    platform_globs = [
+      "ios/**",
+      "{modules,source}/**/*.{h,m,mm}",
```
```suggestion
      '{modules,source}/**/*.{h,m,mm}',
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+    ]
+  end
+
+  native_file_changed = (source_globs + platform_globs).any? {
```
For multi-line blocks, you should use `do...end` instead of `{}`.  (They are identical in this case.)

Also, I'd put `|glob|` either on its own line or after `do`.

e.g.
```ruby
(source_globs + platform_globs).any? do |glob|
  [...]
end
```

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,76 @@ def travis?
+end
+
+# checks if the native build needs to be run
+def should_build?
```
RuboCop wants this to get split up a bit; it's a bit long, and the ABC and Cyclomatic complexity are a bit high. But this is a very good start.

hawkrives replied:
I pulled the big block in the middle out into a separate function; hopefully that helps

rye replied:
Those metrics are still high, but more acceptable this time around.

hawkrives committed:
ensure build-status file exists

rye reviewed (changes requested):
General feedback, mostly. I'd love to have the RuboCop-y things (especially single quotes over double quotes) get taken care of.

hawkrives committed:
split out large block of code into npm_native_package_changed? function

hawkrives committed:
change block into do/end style

hawkrives committed:
single-quotes to double

hawkrives committed:
ensure platform_globs exists

rye committed:
circle: Add a Git history reconciliation step

Since Circle does some really strange stuff when it clones our
repository (resetting the default branch to the branch we're testing,
thereby making it impossible to actually tell what's changed from our
default branch) we need to do some magic to get back on track.

The last reset might not be necessary.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
Tested-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives committed:
quote circle branch vars in bash-land

rye committed:
CHANGELOG: Summarize changes in this PR

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Merge branch 'master' into skip-native-builds

hawkrives committed:
pull out should_build? into more methods

rye reviewed (dismissed):
One potential issue I still see is that `git diff` will also include forwards changes that we don't currently have in our branch.

For instance, if `master` has additional changes to native code, that'll trigger native builds in a JS-only branch, because the diff expands to include stuff that we're missing.

We should use `git log` instead of `git diff`.

I think that the following pipeline satisfies what we need:

```
git log --name-only --format='' 'master'..'<branch>' | sort | uniq
```

- The `..` is required here
- `sort` and `uniq` can be replaced by pipeline magic in Ruby.

This would also give us the opportunity to check, say, how _many_ times a file was changed.

hawkrives committed:
fix missing `return`

hawkrives committed:
address @rye's git feedback

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,86 @@ def travis?
+  changed_dependencies = deps_diff(old_package, new_package, ['dependencies']).keys
+  changed_devdependencies = deps_diff(old_package, new_package, ['devDependencies']).keys
+
+  dep_name_regex = /react|jsc/
```
Can we pull this out into a constant and freeze it?

À la `DEP_NAME_REGEXP = /react|jsc/.freeze` outside this method, and change your comparison to `dep =~ DEP_NAME_REGEXP`.

rye commented on fastlane/lib/util.rb:
```diff
@@ -45,3 +45,86 @@ def travis?
+
+  # 2. check for "packages we care about"
+  if changed_files.include? 'package.json' and npm_native_package_changed?
+    UI.message('should_build? some react|jsc dep changed, so yes')
```
I ask about extracting `DEP_NAME_REGEXP` so that we can include that here:

Once you have it extracted, this should work:

```suggestion
    UI.message("should_build? some dependency matching #{DEP_NAME_REGEXP.inspect} changed, so yes")
```

hawkrives replied:
yeah just wasn't sure how to include it in the string before

rye reviewed (commented):
One more suggestion and then I'm ready to ship.

hawkrives committed:
pull out dep_name_regex and include it in the "should_build?" messages

rye reviewed (approved):
Looks great. :shipit: 

Thanks for taking this on, @hawkrives! 🎉

Side note: At some point I want to start testing this with full mocking just so we don't have any guessing games about logic or stuff like that.  But this was a good expedition.
